### ISSUE TITLE
*Feature Request* Use fields in Populate Virtuals

### ISSUE THREAD

**Issue by florianbepunkt** [Automattic/mongoose#5704]
It would be useful for some use cases to set the references in populated virtuals based on existing fields in the doc, if this is possible. See the following example. If the field `internalOrderId` in the Refund document is no null, it uses `sourceOrderId` as localField, if it is null, it uses `_id`.

You can do something similar already when you populate fields using a dynamic ref, but since I can't use the _id field as reference this is not an option. As far as I know, you can't do the same when you use populate virtuals.

```
RefundSchema.virtual('order', {
  ref: 'Order',
  localField: (this.internalOrderId !== null) ? 'sourceOrderId' : 'internalOrderId',
  foreignField: (this.internalOrderId !== null) ? 'sourceId' : '_id',
  justOne: false
});
```

**Comment by vkarpov15** [comment]
I like this idea, should be doable, will investigate for next minor release.

**Comment by florianbepunkt** [comment]
@vkarpov15 We just updated our mongoose version and I tried to get this to work. Could you please add the syntax to the documentation or provide a basic example?

I have this but it throws an error `TypeError: Invalid `path`. Must be either string or array`

Baiscally what I try to accomplish here is to set the virtual dynamically based on a field of the source document (RefundSchema)

```
RefundSchema.virtual('order', {
  ref: 'Order',
  localField: function() {
    if (this.internalOrderId) {
      return this.internalOrderId;
    } else {
      return this.sourceOrderId;
    }
  },
  foreignField: function() {
    if (this.internalOrderId) {
      return this._id;
    } else {
      return this.sourceId;
    }
  },
  justOne: function() {
    if (this.internalOrderId) {
      return true;
    } else {
      return false;
    }
  },
});
```

**Comment by florianbepunkt** [comment]
Okay, I guess I figured out the syntax, but there seems to be a bug…

RefundSchema.virtual('order', {
  ref: 'Order',
  localField: function() {
    if (this.internalOrderId) {
      // CASE A
      return 'internalOrderId';
    } else {
      // CASE B
      return 'sourceOrderId';
    }
  },
  foreignField: function() {
    if (this.internalOrderId) {
      // CASE A
      return '_id';
    } else {
      // CASE B
      return 'sourceId';
    }
  },
  justOne: false,
});

**Comment by florianbepunkt** [comment]
Okay, I figured out the syntax but there seems to be a bug.... I have the following virtual with dynamic localField and foreignField

```
RefundSchema.virtual('order', {
  ref: 'Order',
  localField: function() {
    if (this.internalOrderId) {
      // CASE A
      return 'internalOrderId';
    } else {
      // CASE B
      return 'sourceOrderId';
    }
  },
  foreignField: function() {
    if (this.internalOrderId) {
      // CASE A
      return '_id';
    } else {
      // CASE B
      return 'sourceId';
    }
  },
  justOne: false,
});
```
Now I have two queries that produce inconsitent results:

```
// QUERY A - returns doc with _id 5a317e080732e735527911fe whic is CASE A for sure
Refund.find({_id: '5a317e080732e735527911fe'}).populate('order').exec().then((resp) => {
  console.log(JSON.stringify(resp, null, 3));
})

// QUERY B - should return CASE A docs and CASE b docs
Refund.find({}).populate('order').exec().then((resp) => {
  console.log(JSON.stringify(resp, null, 3));
})
```
**Actual Behaviour**
Query A does populate the `order` field. However Query B returns (amongst others) the document form Query A, but does not populate the `order` field

**Expected Behaviour**
Both queries should populate the order field for document 5a317e080732e735527911fe

**Comment by vkarpov15** [comment]
Thanks for reporting @florianbepunkt , will investigate

**Comment by vkarpov15** [comment]
Thanks for your patience @florianbepunkt , we really kicked this issue very far down the road. The issue should be fixed in 5.2.18.
